engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java

```diff
@@ -214,19 +217,35 @@ private FunctionCallInfo functionCallInfoById(UUID nodeId) {
     return calls.get(nodeId);
   }
 
-  private String[] typeOf(Object value) {
+  private TypeInfo typeOf(Object value) {
     if (value instanceof UnresolvedSymbol) {
-      return new String[] {Constants.UNRESOLVED_SYMBOL};
+      return TypeInfo.ofType(Constants.UNRESOLVED_SYMBOL);
     }
 
-    var typeOfNode = TypeOfNode.getUncached();
-    Type[] allTypes = value == null ? null : typeOfNode.findAllTypesOrNull(value, true);
-    if (allTypes != null) {
-      String[] result = new String[allTypes.length];
-      for (var i = 0; i < allTypes.length; i++) {
-        result[i] = getTypeQualifiedName(allTypes[i]);
+    if (value != null) {
+      final TypeOfNode typeOfNode = TypeOfNode.getUncached();
+      final Type[] publicTypes = typeOfNode.findAllTypesOrNull(value, false);
+
+      if (publicTypes != null) {
+        final Type[] allTypes = typeOfNode.findAllTypesOrNull(value, true);
+        // Relies on the fact that allTypes appends extra types to the end of publicTypes.
+        assert (publicTypes.length == 0
+            || Objects.equals(
+                publicTypes[publicTypes.length - 1], allTypes[publicTypes.length - 1]));
```

Comment:
This is checking n-th element, not a subrange. Following might be better:
```suggestion
                Arrays.asList(publicTypes).subList(0, publicTypes.length).equals(Arrays.asList(allTypes).subList(0,publicTypes.length));
```

Reply:
did it like that in the end
```java
assert Arrays.equals(publicTypes, Arrays.copyOfRange(allTypes, 0, publicTypes.length));
```